### DISCUSSION TITLE
Programming a CLI-midi controller with python and mido

### DISCUSSION THREAD

**Discussion by ahnwar14** [mido/mido#468] [Q&A]
Hi, im looking for someone to lead me in the right direction with an issue im having. im trying to program a python script which can be used as a cli midi controller and eliminate the need for external hardware or software. Based on user input, it can toggle effects states or change amp sim parameters. The current issue is after having created the virtual ports (visible in patchage), im not sure how i would go about connecting them to the software applications running the effects (such as MODEP or Guitarix, both running on patchbox os)

below is current code i written used on the Rpi 3b+. Ive also asked around in the [blokas forums](https://community.blokas.io/t/toggling-effects-with-python-script/4359/1) .

```python
#!/usr/bin/python
import mido

# these ports show up in patchage
virtualMidiIn = mido.open_input('vpIn', virtual=True) 
virtualMidiOut = mido.open_output('vpOut', virtual=True)


# main menu
while True:
    menuChar = input("Enter Menu Selection: ")

    if "quit" in menuChar:
        print("quitting program")
        virtualMidiIn.close()
        virtualMidiOut.close()
        print(mido.get_ioport_names())
        print(mido.get_input_names())
        print(mido.get_output_names())
        break
    else:
        if "dst" in menuChar:
            print("this turns on distortion")
        elif "rvb" in menuChar:
            print("This toggles reverb")
        elif "phs" in menuChar:
            print("This toggles phase")
        elif "dly" in menuChar:
            print("this toggles delay")
        elif "status" in menuChar:
            print(mido.get_ioport_names())
            print(mido.get_input_names())
            print(mido.get_output_names())
        else:
            print("Please enter valid choice")
```

**Comment by rdoursenaud** [comment]
Hi,

AFAIK Connecting two MIDI ports together is not an operation that `mido` can do on its own.
Reading the thread you linked from the forum, I gathered that you're probably using ALSA MIDI.
From there, you have multiple options:
- Use the ALSA's `aconnect` CLI utility and, to automate the process, call it from your code or use Python ALSA bindings.
- Assuming the target software is already running, you could simply use `open_input` and `open_output` on the application's ports instead of creating virtual ports.

P.S.: I have edited your message slightly to properly highlight the code snippet.

**Reply by ahnwar14** [reply]
Hey, i was able to create two virtual ports and connect them and send messages to the virtual input using the virtual output. Using `a2jmidid-e` , im able to convert them to jack. Do you know if mido also works with jack midi?

**Reply by rdoursenaud** [reply]
`mido` doesn't have a JACK backend that I'm aware of at the moment.

`rtmidi`, our current default backend, seem to be able to use JACK but this requires recompilation (or distribution support in the form of a specific package) and I haven't tested this yet.

We have a plan to try and bring native backends at some point and JACK seems to be a good candidate. Nothing materialized at the moment unfortunately.

Look like you'll have to use some kind of bridge for now (`a2jmidid` maybe).
